### PR TITLE
fix(catalog-backend-module-gitlab): #20891 GitlabOrgDiscoveryEntityProvider inherit members users

### PR DESCRIPTION
**Pull request by pushit-tech**
… gets all inherited users when consulting /groups/id/members

## Hey, I just made a Pull Request!

See the error for more context #20891 

I added an `/all` to the `/gorups/id/members` API endpoint of `catalog-backend-module-gitlab`
according to [gitlab docs](https://docs.gitlab.com/ee/api/members.html#list-all-members-of-a-group-or-project-including-inherited-and-invited-members) 

#### :heavy_check_mark: Checklist

- [x] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [ ] Added or updated documentation
- [ ] Tests for new functionality and regression tests for bug fixes
- [ ] Screenshots attached (for UI changes)
- [x] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

### CODE REVIEW

**Comment by matteosilv** [comment]
looks good. However I still don't understand the use case. I assume you have a SaaS subscription. Why should you have members only on subgroups? 
Anyway this doesn't impact people who doesn't, so I guess it's ok

**Comment by pushit-tech** [comment]
Thank you for your input, @matteosilv . Yes, we do utilize a SaaS subscription. The decision to manage users solely within subgroups, rather than in the main organization group, is based on internal security policies.

Will be really usefull to have the inherited group members on the discovery 🚀, I will be open to any suggestion to have this change merged 

Thanks

**Comment by benjdlambert** [comment]
@matteosilv @pushit-tech so, just so we're clear here, this endpoint works without any SaSS subscription and shouldn't impact anyone else? Just trying to understand how risky this is.

**Comment by pushit-tech** [comment]
Hi @benjdlambert,

Yes, you're correct.

In the master branch, we utilize the same [members API](https://docs.gitlab.com/ee/api/members.html#list-all-members-of-a-group-or-project-including-inherited-and-invited-members). With this Merge Request (MR), we extend the functionality to retrieve inherited users, which, as per the documentation, isn't tied to a specific subscription.

The logic for this endpoint is situated within the else block of an isSelfManaged validation, indicating it will impact only the SaaS subscription, as shown below:

```javascript 
if (client.isSelfManaged()) {
  //self manage gitlab code here
} else {
 //SaaS logic code here, where we use the listSaaSUsers and /members/all endpoint
}
```
Although there's a Premium attribute within [this endpoint](https://docs.gitlab.com/ee/api/members.html#list-all-members-of-a-group-or-project-including-inherited-and-invited-members) related to the state, we are not utilizing it.

image

**Comment by pushit-tech** [comment]
Hi guys is there anything pending here? 

Should I make any update ? 

Will be grate to have this on the closest release.

**Comment by freben** [comment]
The maintainers are off this week. We'll get back to ya

**Comment by matteosilv** [comment]
Since we have to wait for review. It's ok if I add I change I need to derive the user email from the username if email is not readable?

**Comment by pushit-tech** [comment]
@matteosilv and does it always apply that the user email is the same as the user name ? 

Or can be this a note on the docs, establishing the requirements so that we can get the email ?

**Comment by matteosilv** [comment]
Our use case is:
- we create user in GitLab through SCIM gitlab integration. Users are created by taking the name part of the email and adding numbers if the user already exists on gitlab (that's GitLab's own workflow)
- In GitLab saas the members/all api only shows emails when the user is Enterprise (so it's fully managed by gitlab, check out their doc if you want to know more).
- For the reason above, for many of our users we CANNOT read the email, but we can derive it from the username, by just removing the numbers at the end.
- I implemented the logic above via backstage post-processor


We cannot do much to retrieve the email from gitlab directly as it is their own limitation

**Review by nuritizra — APPROVED** [review]

**Review by benjdlambert — APPROVED** [review]
Thanks :tada:

**Pull request merged**
